File: python/TEST_event.py
#!/usr/bin/env python
# coding: utf-8
# vim: set sw=4 ts=4 sts=4 et tw=80 :

import os, signal, sys, unittest

sys.path.insert(0, '.')
from pinktrace import event, trace

class TestEvent_01_Invalid(unittest.TestCase):

    def test_01_decide(self):
        self.assertRaises(TypeError, event.decide)
        self.assertRaises(TypeError, event.decide, 'pink')

class TestEvent_02(unittest.TestCase):

    def test_01_event_unknown(self):
        ev = event.decide(-1)
        self.assertEqual(ev, event.EVENT_UNKNOWN)

    def test_02_event_stop(self):
        pid = os.fork()
        if not pid: # child
            trace.me()
            os.kill(os.getpid(), signal.SIGSTOP)

            os._exit(0)
        else: # parent
            pid, status = os.waitpid(pid, 0)
            ev = event.decide(status)
            self.assertEqual(ev, event.EVENT_STOP)

            try: trace.kill(pid)
            except OSError: pass

    def test_03_event_syscall(self):
        pid = os.fork()
        if not pid: # child
            trace.me()
            os.kill(os.getpid(), signal.SIGSTOP)

            os.sleep(1)
            os._exit(0)
        else: # parent
            os.waitpid(pid, 0)
            trace.setup(pid)

            trace.syscall(pid)
            pid, status = os.waitpid(pid, 0)
            ev = event.decide(status)
            self.assertEqual(ev, event.EVENT_SYSCALL)

            try: trace.kill(pid)
            except OSError: pass

    def test_04_event_fork(self):
        pid = os.fork()
        if not pid: # child
            trace.me()
            os.kill(os.getpid(), signal.SIGSTOP)

            os.fork()
            os._exit(0)
        else: # parent
            os.waitpid(pid, 0)
            trace.setup(pid, trace.OPTION_SYSGOOD | trace.OPTION_FORK)

            trace.cont(pid)
            pid, status = os.waitpid(pid, 0)
            ev = event.decide(status)
            self.assertEqual(ev, event.EVENT_FORK)

            try: trace.kill(pid)
            except OSError: pass

    def test_05_event_vfork(self):
        pass

    def test_06_event_clone(self):
        pass

    def test_07_event_vfork_done(self):
        pass

    def test_08_event_exec(self):
        pid = os.fork()
        if not pid: # child
            trace.me()
            os.kill(os.getpid(), signal.SIGSTOP)

            os.execvp("/bin/true", ("true",))
            os._exit(0)
        else: # parent
            os.waitpid(pid, 0)
            trace.setup(pid, trace.OPTION_SYSGOOD | trace.OPTION_EXEC)

            trace.cont(pid)
            pid, status = os.waitpid(pid, 0)
            ev = event.decide(status)
            self.assertEqual(ev, event.EVENT_EXEC)

            try: trace.kill(pid)
            except OSError: pass

    def test_09_event_exit(self):
        pid = os.fork()
        if not pid: # child
            trace.me()
            os.kill(os.getpid(), signal.SIGSTOP)

            os._exit(13)
        else: # parent
            os.waitpid(pid, 0)
            trace.setup(pid, trace.OPTION_SYSGOOD | trace.OPTION_EXIT)

            trace.cont(pid)
            pid, status = os.waitpid(pid, 0)
            ev = event.decide(status)
            self.assertEqual(ev, event.EVENT_EXIT)
            msg = trace.geteventmsg(pid)
            self.assertEqual(os.WEXITSTATUS(msg), 13)

            try: trace.kill(pid)
            except OSError: pass

    def test_10_event_genuine(self):
        pid = os.fork()
        if not pid: # child
            trace.me()
            os.kill(os.getpid(), signal.SIGSTOP)

            os.kill(os.getpid(), signal.SIGTSTP)
            os._exit(0)
        else: # parent
            os.waitpid(pid, 0)
            trace.setup(pid)

            trace.cont(pid)
            pid, status = os.waitpid(pid, 0)
            ev = event.decide(status)
            self.assertEqual(ev, event.EVENT_GENUINE)

            try: trace.kill(pid)
            except OSError: pass

    def test_11_event_exit_genuine(self):
        pid = os.fork()
        if not pid: # child
            trace.me()
            os.kill(os.getpid(), signal.SIGSTOP)

            os._exit(0)
        else: # parent
            os.waitpid(pid, 0)
            trace.setup(pid)

            trace.cont(pid)
            pid, status = os.waitpid(pid, 0)
            ev = event.decide(status)
            self.assertEqual(ev, event.EVENT_EXIT_GENUINE)

            try: trace.kill(pid)
            except OSError: pass

    def test_12_event_exit_signal(self):
        pid = os.fork()
        if not pid: # child
            trace.me()
            os.kill(os.getpid(), signal.SIGSTOP)

            os.kill(os.getpid(), signal.SIGKILL)
        else: # parent
            os.waitpid(pid, 0)
            trace.setup(pid)

            trace.cont(pid)
            pid, status = os.waitpid(pid, 0)
            ev = event.decide(status)
            self.assertEqual(ev, event.EVENT_EXIT_SIGNAL)

            try: trace.kill(pid)
            except OSError: pass

if __name__ == '__main__':
    unittest.main()

File: ruby/TEST_event.rb
#!/usr/bin/env ruby
# coding: utf-8
# vim: set sw=2 sts=2 et nowrap fenc=utf-8 :

require 'test/unit'

$:.insert(0, '.libs')
require 'PinkTrace'

class TestPinkEvent < Test::Unit::TestCase
  def test_event_unknown
    event = PinkTrace::Event.decide -1
    assert(event == PinkTrace::Event::EVENT_UNKNOWN, "Wrong event, expected: UNKNOWN got: #{event}")
  end

  def test_event_stop
    pid = fork do
      PinkTrace::Trace.me
      Process.kill 'STOP', Process.pid
    end
    Process.waitpid pid
    event = PinkTrace::Event.decide
    assert(event == PinkTrace::Event::EVENT_STOP, "Wrong event, expected: STOP got: #{event}")

    begin PinkTrace::Trace.kill pid
    rescue Errno::ESRCH ;end
  end

  def test_event_syscall
    pid = fork do
      PinkTrace::Trace.me
      Process.kill 'STOP', Process.pid

      sleep 1
    end
    Process.waitpid pid
    PinkTrace::Trace.setup pid

    PinkTrace::Trace.syscall pid
    Process.waitpid pid
    event = PinkTrace::Event.decide
    assert(event == PinkTrace::Event::EVENT_SYSCALL, "Wrong event, expected: SYSCALL got: #{event}")

    begin PinkTrace::Trace.kill pid
    rescue Errno::ESRCH ;end
  end

  def test_event_fork
    pid = fork do
      PinkTrace::Trace.me
      Process.kill 'STOP', Process.pid

      fork { exit 0 }
    end
    Process.waitpid pid
    PinkTrace::Trace.setup pid, (PinkTrace::Trace::OPTION_SYSGOOD | PinkTrace::Trace::OPTION_FORK)

    PinkTrace::Trace.cont pid
    Process.waitpid pid
    event = PinkTrace::Event.decide
    assert(event == PinkTrace::Event::EVENT_FORK, "Wrong event, expected: FORK got: #{event}")

    begin
      child = PinkTrace::Trace.geteventmsg pid
      PinkTrace::Trace.kill child
      PinkTrace::Trace.kill pid
    rescue Errno::ESRCH ;end
  end

  def test_event_vfork
  end

  def test_event_clone
  end

  def test_event_vfork_done
  end

  def test_event_exec
    pid = fork do
      PinkTrace::Trace.me
      Process.kill 'STOP', Process.pid

      exec '/bin/true'
    end
    Process.waitpid pid
    PinkTrace::Trace.setup pid, (PinkTrace::Trace::OPTION_SYSGOOD | PinkTrace::Trace::OPTION_EXEC)

    PinkTrace::Trace.cont pid
    Process.waitpid pid
    event = PinkTrace::Event.decide
    assert(event == PinkTrace::Event::EVENT_EXEC, "Wrong event, expected: EXEC got: #{event}")

    begin PinkTrace::Trace.kill pid
    rescue Errno::ESRCH ;end
  end

  def test_event_exit
    pid = fork do
      PinkTrace::Trace.me
      Process.kill 'STOP', Process.pid

      exit 0
    end
    Process.waitpid pid
    PinkTrace::Trace.setup pid, (PinkTrace::Trace::OPTION_SYSGOOD | PinkTrace::Trace::OPTION_EXIT)

    PinkTrace::Trace.cont pid
    Process.waitpid pid
    event = PinkTrace::Event.decide
    assert(event == PinkTrace::Event::EVENT_EXIT, "Wrong event, expected: EXIT got: #{event}")

    begin PinkTrace::Trace.kill pid
    rescue Errno::ESRCH ;end
  end

  def test_event_genuine
    pid = fork do
      PinkTrace::Trace.me
      Process.kill 'STOP', Process.pid

      Process.kill 'TSTP', Process.pid
    end
    Process.waitpid pid
    PinkTrace::Trace.setup pid

    PinkTrace::Trace.cont pid
    Process.waitpid pid
    event = PinkTrace::Event.decide
    assert(event == PinkTrace::Event::EVENT_GENUINE, "Wrong event, expected: GENUINE got: #{event}")

    begin PinkTrace::Trace.kill pid
    rescue Errno::ESRCH ;end
  end

  def test_event_exit_genuine
    pid = fork do
      PinkTrace::Trace.me
      Process.kill 'STOP', Process.pid

      exit 0
    end
    Process.waitpid pid
    PinkTrace::Trace.setup pid

    PinkTrace::Trace.cont pid
    Process.waitpid pid
    event = PinkTrace::Event.decide
    assert(event == PinkTrace::Event::EVENT_EXIT_GENUINE, "Wrong event, expected: EXIT_GENUINE got: #{event}")
    begin PinkTrace::Trace.kill pid
    rescue Errno::ESRCH ;end
  end

  def test_event_exit_signal
    pid = fork do
      PinkTrace::Trace.me
      Process.kill 'STOP', Process.pid

      Process.kill 'KILL', Process.pid
    end
    Process.waitpid pid
    PinkTrace::Trace.setup pid

    PinkTrace::Trace.cont pid
    Process.waitpid pid
    event = PinkTrace::Event.decide
    assert(event == PinkTrace::Event::EVENT_EXIT_SIGNAL, "Wrong event, expected: EXIT_SIGNAL got: #{event}")
    begin PinkTrace::Trace.kill pid
    rescue Errno::ESRCH ;end
  end
end
